src/com/photoncat/aiproj2/heuristic/Heuristic2.java

```diff
@@ -0,0 +1,902 @@
+            return -50000000;
+        }
+        else {
+            int sum=0;
+            int N=board.getSize();
+            for(int i=0;i<N;i++){
+                for (int j=0;j<N;j++){
+                    if (board.getPiece(i,j)==PieceType.NONE){
+                        sum=sum+evaluateNByM(board,i,j,PieceType.CROSS);
+                        sum=sum-evaluateNByM(board,i,j,PieceType.CIRCLE);
+                    }
+
+                }
+            }
+            return sum;
+        }
+    }
+    private int NByFive(Board board) {
+        if (checkWin(board,PieceType.CROSS)){
+            return 10000000;
+        }
+        else if (checkWin(board,PieceType.CIRCLE)){
+            return -10000000;
+        }
+        else {
+            int sum=0;
+            int N=board.getSize();
+            for(int i=0;i<N;i++){
+                for (int j=0;j<N;j++){
+                    if (board.getPiece(i,j)==PieceType.NONE){
+//                         sum=sum+evaluateNByFive(board,i,j,PieceType.CROSS);
+//                         sum=sum-evaluateNByFive(board,i,j,PieceType.CIRCLE);
+//                         sum=sum+evaluateNByFive2(board,i,j,PieceType.CROSS);
+//                         sum=sum-evaluateNByFive2(board,i,j,PieceType.CIRCLE);
+//                         sum=sum+evaluateNByM(board,i,j,PieceType.CROSS);
+//                         sum=sum-evaluateNByM(board,i,j,PieceType.CIRCLE);
+                        sum=sum+evaluateNByFive3(board,i,j,PieceType.CROSS);
+                        sum=sum-evaluateNByFive3(board,i,j,PieceType.CIRCLE);
+                    }
+
+                }
+            }
+            return sum;
+        }
+    }
+    private int NByFour(Board board) {
+        if (checkWin(board,PieceType.CROSS)){
+            return 50000000;
+        }
+        else if (checkWin(board,PieceType.CIRCLE)){
+            return -50000000;
+        }
+        else {
+            int sum=0;
+            int N=board.getSize();
+            for(int i=0;i<N;i++){
+                for (int j=0;j<N;j++){
+                    if (board.getPiece(i,j)==PieceType.NONE){
+                        sum=sum+evaluateNByFour(board,i,j,PieceType.CROSS);
+                        sum=sum-evaluateNByFour(board,i,j,PieceType.CIRCLE);
+                    }
+
+                }
+            }
+            return sum;
+        }
+    }
+
+    private int evaluateNByM(Board board, int row, int col, PieceType player) {
+        int M=board.getM();
+        int score=0;
+        int cur=0;
+        int opp=0;
+        if (player==PieceType.CROSS){
+            cur=1;
+            opp=2;
+        }
+        else {
+            cur=2;
+            opp=1;
+        }
+        int numOfTwo=0;
+        for (int i=1;i<=8;i++){
+            //top level--------------------------------------
+            //01111* or 21111*
+            int idx=-1;
+            boolean flagFirst=true;
+            for (;idx>-M;idx--){
```

Comment:
Consider using while loops for those who don't have an initializing variable needed.